### DISCUSSION TITLE
Conversion to and from openxlsx

### DISCUSSION THREAD

**Discussion by Arthfael** [JanMarvin/openxlsx2#948] [Ideas]
I have read the openxlsx2 documentation and understand the rationale for essentially re-tackling the problem openxlsx aimed to solve using a fresh approach. I have encountered some of the limitations of the otherwise excellent openxlsx, so am looking at transitioning to openxlsx2. _(Specifically, I want to write text with several colours in a cell, which I think I can achieve with fmt_txt())_
However, despite the much appreciated documentation on functions equivalence... that is a lot of code to rewrite/troubleshoot (especially because I made a lot of use of the Styles objects which appear not to have equivalents in openxlsx2). I am using it to write Excel report tables at the end of a lengthy data analysis. As a short/middle term solution, to ease adoption of openxlsx2, it would be convenient to be able to first write R data into an openxlsx workbook, then convert the object to an openxlsx2 workbook so as to perform more advanced tasks, before saving it locally as xlsx. At least in this direction it should make sense.
So the question I would respectfully like ask is: would it even make sense to create a converter between both packages' workbook objects? If so, is this something which is already in the works?
(Of course, an alternative path would be for me to save the openxlsx workbook to disc, then load it with openxlsx2; this may not be the most efficient though.)
Thoughts on this?

**Comment by JanMarvin** [comment]
Hi @Arthfael , thanks for the suggestion. From a users point of view it is very understandable. Just yesterday evening I was chatting with a friend about R, LaTeX and stable software and my hesitation to change anything on my computer. And I do not switch from one package to another, because sometimes things need time to settle.
Ofc it should be possible to convert `openxlsx` into `openxlsx2`, but it will not be a task for a rainy afternoon I'm afraid. And not one that I have any real need for. Therefore, if someone wants to do this, please go ahead.

As you wrote, in many cases it might simply be enough to convert via the output xlsx file.

```R
tempfile <- openxlsx2::temp_xlsx()
openxlsx::saveWorkbook(wb_openxlsx, tempfile)
wb_openxlsx2 <- openxlsx2::wb_load(tempfile)
```

And for many others it might simply be enough to rewrite some older code from scratch. But we've made this a new package, so that users still have the chance to use `openxlsx` as long as they need/want. And even though `openxlsx2` has seen a lot of development since the fork, `openxlsx` still gets infrequent updates (but I haven't been in contact with Philipp for at least two years I would say).

**Reply by Arthfael** [reply]
Thank you very much Jan, having written my own converters for much easier data structures I do agree that it would be no simple task ^^
I did try my own suggestion of doing parts of the work in openxlsx and parts in openxlsx2, but encountered some issues with embedded images. So now I am trying to rewrite my old code in openxlsx2... many functions are an easy translation, although you will not be surprised to hear that the transition from openxlsx style objects to the openxlsx2 way of dealing with style is proving... challenging. I will persevere though, because I do need a better solution for applying different colour styles within a single cell than my current one *, and your package is the only R-contained solution for doing this I know of.

(* an Excel macro which is run by a shell call from R, written by my former technician who left years ago... this is clunky and will break eventually)

**Reply by JanMarvin** [reply]
Have a look at [the upgrade](https://janmarvin.github.io/ox2-book/chapters/openxlsx2_upgrade_from_openxlsx.html) and [the styling manual](https://janmarvin.github.io/ox2-book/chapters/openxlsx2_style_manual.html) and feel free to ask if you have any problems, that's what this discussion board is for. I don't have the time (and tbh interest 😉 ) to rewrite all your scripts, but after all, I'm probably the person who knows the inner details of `openxlsx2` best. So if you have a small reproducible chunk and want to know how to convert it, maybe I can help you.

It's not so much the complexity of the task, but I just don't really have a need for it. In the time it would take me to write this, I could be reading a book, watching TV or going to the pub. Things that normal people do in their free time :)

**Comment by Arthfael** [comment]
Dear Jan,

I perfectly understand - as I said, it was just an idea. And far from me
the thought to ask you to rewrite my scripts!
(Now, exploring the alternative "openxlsx->excel->openxlsx2" solution , I
am encountering some issues which I will aim to report with reprex as soon
as I start getting a grip on what is happening.)
Thanks for your help!
Kind regards,

Armel

**Comment by Arthfael** [comment]
So, my investigations quite clearly show that the errors are with openxlsx,
not openxlsx2, so I will report there ^^ Only one of the issues I have
investigated so far came solely from openxlsx2: in my hands with the latest
version the start_row/start_col arguments are ignored for ...$add_image()
even if dims is not provided.

**Reply by JanMarvin** [reply]
Iirc we don’t support `from_{col,row}` anymore. Unless you see a deprecation warning, we only support `dims`. If you see a deprecation warning, we support it only as a backwards compatible fallback option that will be removed in the future.

Reporting issues to `openxlsx` is currently as effective as screaming in the woods. I was among the last to fix issues and due to my own time constraints, I’ve stopped working on it a while ago.

**Reply by JanMarvin** [reply]
Looks like you found a bug. We never convert to `dims` in `wb_add_image()`:

https://github.com/JanMarvin/openxlsx2/blob/ee3590b2ac30dd17a646ccef450aa8fb7399d514/R/class-workbook.R#L4915-L4922

Should be a simple fix using dims and `wb_dims()`.

**Reply by JanMarvin** [reply]
In the development branch both functions below now behave identical:

``` r
library(openxlsx2)
img <- system.file("extdata", "einstein.jpg", package = "openxlsx2")

wb <- wb_workbook()$add_worksheet()$add_image(file = img, dims = wb_dims(rows = 5, cols = 5))

wb <- wb_workbook()$add_worksheet()$add_image(file = img, start_row = 5, start_col = 5)
#> Warning: 'start_col/start_row' is deprecated.
#> Use 'dims' instead.
#> See help("Deprecated") and help("openxlsx2-deprecated").
```

**Comment by Arthfael** [comment]
Good to know. I had already reported some and saw no action but... oh well.
I am rewriting everything using openxlsx2 now. And dims, which is quite
easy with wb_dims().

For the record, I think I found an efficient, practically trouble-free
solution to trick openxlsx2 into using my openxlsx styles - these were so
many, and so complex that I just couldn't face the prospect of spending
days re-writing them. Instead, I am:

   - using openxlsx to create a dummy xlsx with a single tab containing one
   cell for each old openxlsx style, then
   - loading that .xlsx with openxlsx2 and using wb_get_cell_style() to
   propagate the style from each style cell, then removing the dummy tab

This seems to work, with only a warning in R when I run openxlsx2::wb_save
- but the .xlsx file then opens in Excel without needing repairs.

**Reply by JanMarvin** [reply]
Well, to put it kindly, you've made it pretty clear what you think about the handling of cell styles in `openxlsx2`. Since I have no idea of the complexity of your code and you haven't shared anything to look at, you should do what works best for you :)

What you describe here looks like a pretty pragmatic way of handling it. You could even modify the template with Excel, which should load in `openxlsx2` without warning (I assume the warning indicates that something is going slightly wrong in `openxlsx`).

**Comment by Arthfael** [comment]
Actually, I do not dislike it. There are probably very good reasons to do
it this way. The only thing which is tricky is the conversion of an R
script relying heavily on a complex collection of openxlsx Style objects to
openxlsx2 - possibly the only thing hampering transition from the older
package.
Anyway, with the help of the dummy tab I just successfully completed my
conversion, thanks for your support, I am looking forward to exploring the
possibilities of your package further! Also very good point about how to
fix the warning by opening/saving with Excel - especially since the warning
did decay into full blown error once I started embedding images (could not
save anymore), so thanks.
